near/nep141-locker/src/lib.rs

```diff
@@ -549,9 +587,94 @@ impl Contract {
         }
     }
+
+    fn insert_raw_transfer(
+        &mut self,
+        nonce: u128,
+        transfer_message: &TransferMessage,
+        sender_id: &AccountId,
+    ) -> Option<Vec<u8>> {
+        self.pending_transfers.insert_raw(
+            &borsh::to_vec(&nonce).sdk_expect("ERR_BORSH"),
+            &TransferMessageStorage::encode_borsh(transfer_message, sender_id)
+                .sdk_expect("ERR_BORSH"),
+        )
+    }
+
+    fn add_transfer_message(
+        &mut self,
+        nonce: u128,
+        transfer_message: &TransferMessage,
+        message_owner: &AccountId,
+    ) -> NearToken {
+        let storage_usage = env::storage_usage();
+        self.insert_raw_transfer(nonce, transfer_message, message_owner);
+        env::storage_byte_cost().saturating_mul((env::storage_usage() - storage_usage).into())
+    }
+
+    fn remove_transfer_message(&mut self, nonce: u128) -> TransferMessage {
+        let storage_usage = env::storage_usage();
+        let (transfer_message, owner) = self
+            .pending_transfers
+            .remove(&nonce)
+            .map(|m| m.into_main())
+            .sdk_expect("ERR_TRANSFER_NOT_EXIST");
+
+        let refund =
+            env::storage_byte_cost().saturating_mul((storage_usage - env::storage_usage()).into());
+
+        let mut storage = self
+            .accounts_balances
+            .get(&owner)
+            .sdk_expect("ERR_ACCOUNT_NOT_REGISTERED");
+
+        storage.available = storage.available.saturating_sub(refund);
+        self.accounts_balances.insert(&owner, &storage);
+
+        transfer_message
+    }
+
+    fn add_fin_transfer(&mut self, chain_kind: ChainKind, nonce: u128) -> NearToken {
+        let storage_usage = env::storage_usage();
+        require!(
+            self.finalised_transfers.insert(&(chain_kind, nonce)),
+            "The transfer is already finalised"
+        );
+        env::storage_byte_cost().saturating_mul((env::storage_usage() - storage_usage).into())
+    }
+
+    fn update_storage_balance(
+        &mut self,
+        account_id: AccountId,
+        required_balance: NearToken,
+        attached_deposit: NearToken,
+    ) {
+        let refund = match self.accounts_balances.get(&account_id) {
+            Some(mut storage) => {
+                if storage.available >= required_balance {
+                    storage.available = storage.available.saturating_sub(required_balance);
+                    self.accounts_balances.insert(&account_id, &storage);
+                    Some(attached_deposit)
+                } else {
+                    None
+                }
+            }
+            None => None,
+        }
+        .unwrap_or_else(|| {
+            attached_deposit
+                .checked_sub(required_balance)
+                .sdk_expect("Not enough storage deposited")
+        });
+
+        if !refund.is_zero() {
+            Promise::new(account_id).transfer(refund);
+        }
```

Comment:
And why is the logic like this? In my mind, I imagined subtracting from the attached funds first and then covering the remaining amount from the storage balance if needed.

Reply:
Like this deab232ed0f853ca79d8ffbdd82924ccb83681c3?